### DISCUSSION TITLE
Other remotes providers

### DISCUSSION THREAD

**Discussion by gsvarovsky** [m-ld/m-ld-spec#91] [Discussion]
### what

Support for additional remotes providers. Comments below with alternative providers.

### why

Fully a peer-to-peer remotes would demonstrate how m-ld can be used in a fully-decentralised deployment.

**Comment by janthecollaborator** [comment]
### Bridgefy

https://bridgefy.me/sdk/

See https://github.com/m-ld/m-ld-spec/discussions/81

- Doesn't even use the internet.
- ❌ Currently free but [will become paid](https://bridgefy.me/faq/) "sometime during 2021".

Potentially some brainstorming around the fact that connectivity between clones will be much more variable, you perhaps don't want Bridgefy to be hopping the message from clone A's messaging layer to clone C's via clone B's, rather, expose to m-ld the lack of connectivity between A and C.

**Comment by gsvarovsky** [comment]
### libp2p

https://libp2p.io/

Status 2022: https://fosdem.org/2022/schedule/event/libp2p/

This is an intermediary library that could allow a freer use of underlying transports.

**Comment by gsvarovsky** [comment]
### XMPP

> XMPP is a open, free and decentralized instant messaging network.

https://xmpp.org

- Decentralised server-based message protocol
- ✅ Browser library support e.g. [JSXC](https://www.jsxc.org/)

Used [by openEngiadina](https://inqlab.net/2021-11-12-openengiadina-from-activitypub-to-xmpp.html) (in preference to ActivityPub and Matrix)

Hosting: https://wiki.xmpp.org/web/XMPP_Hosting_Providers

In particular: https://snikket.org/hosting/

**Comment by gsvarovsky** [comment]
### ØMQ
https://zeromq.org

- ✅ No hosting: decentralised.
- ❌ Websocket transport [exists (JSMQ)](https://github.com/zeromq/JSMQ) but may not be maintained and requires a live server.

**Comment by gsvarovsky** [comment]
### Neuropil
https://www.neuropil.org

(NGI Zero funded)

- ✅ No hosting: decentralised
- ❌ No websocket transport option = no browsers

**Comment by gsvarovsky** [comment]
### Briar
https://briarproject.org/

[Protocol docs](https://code.briarproject.org/briar/briar/-/wikis/A-Quick-Overview-of-the-Protocol-Stack)

- ✅ No hosting: decentralised.
- ❌ No websocket transport option or other browser implementation.

**Comment by gsvarovsky** [comment]
### WAMP

https://wamp-proto.org/
- Publish & Subscribe
- Routed Remote Procedure Calls

Driven and implemented by https://crossbar.io/ (router)
[Native support for Tor](https://blog.torproject.org/secure-messaging-onion-services-how/)

- ✅ Websocket-based, so browser-native clients available
- ❌ Websocket-based, so routers required – and clustering is [paid](https://crossbario.com/)

**Comment by gsvarovsky** [comment]
### p2panda

https://p2panda.org/

> a user-friendly peer-to-peer communications protocol

Based on an append-only log

- ✅ "Browser friendliness"
- ❌ Discovery & Replication ["not implemented"](https://p2panda.org/handbook/docs/category/networking)

May be primarily useful for better journaling and transport security

**Comment by gsvarovsky** [comment]
### WebRTC mesh/gossip

Many papers and packages available ([Google search](https://www.google.com/search?q=webrtc+mesh+gossip&oq=webrtc+mesh+gossip&aqs=chrome..69i57j33i160l2.5243j0j7&sourceid=chrome&ie=UTF-8), [comment in simple-peer library](https://github.com/feross/simple-peer#connecting-more-than-2-peers)), but probably best to roll our own version that suits us.

- ✅ Browser-native (fast)
- ❌ Still requires some server support: discovery, signalling, STUN, TURN
- ❌ Mesh approach will not scale to lots of peers
- ❌ Enterprise networks may not allow it

**Comment by gsvarovsky** [comment]
### Message Passing Interface (MPI)

https://en.wikipedia.org/wiki/Message_Passing_Interface

Standard for high-performance computing. Also note other options in §Future.

May be suitable for HPC scenarios with a native engine.

- ✅ Bare-metal performance
- ⚠️ Agnostic to centralisation (I think) – see e.g. [Decentralized Smart Traffic Framework](https://symposium.foragerone.com/utk-summer-research-scholars-symposium-2022/presentations/47284)
- ❌ No batteries included, need to look into transports, maybe internet support
